code/modules/reagents/withdrawal/_addiction.dm

```diff
@@ -0,0 +1,139 @@
+			if(addiction == type && possible_drug.volume >= addiction_relief_treshold) //If one of them matches, and we have enough of it in our system, we're not losing addiction
+				if(current_addiction_cycle)
+					end_withdrawal(affected_carbon) //stop the pain
+				on_drug_of_this_addiction = TRUE
+				break
+
+	var/withdrawal_stage
+
+	switch(current_addiction_cycle)
+		if(WITHDRAWAL_STAGE1_START_CYCLE to WITHDRAWAL_STAGE1_END_CYCLE)
+			withdrawal_stage = 1
+		if(WITHDRAWAL_STAGE2_START_CYCLE to WITHDRAWAL_STAGE2_END_CYCLE)
+			withdrawal_stage = 2
+		if(WITHDRAWAL_STAGE3_START_CYCLE to INFINITY)
+			withdrawal_stage = 3
+		else
+			withdrawal_stage = 0
+
+	if(!on_drug_of_this_addiction)
+		if(affected_carbon.mind.remove_addiction_points(type, addiction_loss_per_stage[withdrawal_stage + 1] * delta_time)) //If true was returned, we lost the addiction!
+			return
+
+	if(!current_addiction_cycle) //Dont do the effects if were not on drugs
+		return FALSE
+
+	switch(current_addiction_cycle)
+		if(WITHDRAWAL_STAGE1_START_CYCLE)
+			withdrawal_enters_stage_1(affected_carbon)
+		if(WITHDRAWAL_STAGE2_START_CYCLE)
+			withdrawal_enters_stage_2(affected_carbon)
+		if(WITHDRAWAL_STAGE3_START_CYCLE)
+			withdrawal_enters_stage_3(affected_carbon)
+
+	///One cycle is 2 seconds
+	switch(withdrawal_stage)
+		if(1)
+			withdrawal_stage_1_process(affected_carbon, delta_time)
+		if(2)
+			withdrawal_stage_2_process(affected_carbon, delta_time)
+		if(3)
+			withdrawal_stage_3_process(affected_carbon, delta_time)
+
+	LAZYADDASSOC(affected_carbon.mind.active_addictions, type, 1 * delta_time) //Next cycle!
+
+
+//These function very similar to disease stages, so if anyone wants to add a new addiction that does cool things look at how diseases do them
+
+/// Called when addiction enters stage 1
+/datum/addiction/proc/withdrawal_enters_stage_1(mob/living/carbon/affected_carbon)
+	SEND_SIGNAL(affected_carbon, COMSIG_ADD_MOOD_EVENT, "[type]_addiction", light_withdrawal_moodlet, name)
+
+/// Called when addiction enters stage 2
+/datum/addiction/proc/withdrawal_enters_stage_2(mob/living/carbon/affected_carbon)
+	SEND_SIGNAL(affected_carbon, COMSIG_ADD_MOOD_EVENT, "[type]_addiction", medium_withdrawal_moodlet, name)
+
+/// Called when addiction enters stage 3
+/datum/addiction/proc/withdrawal_enters_stage_3(mob/living/carbon/affected_carbon)
+	SEND_SIGNAL(affected_carbon, COMSIG_ADD_MOOD_EVENT, "[type]_addiction", severe_withdrawal_moodlet, name)
+
+/datum/addiction/proc/end_withdrawal(mob/living/carbon/affected_carbon)
+	LAZYSET(affected_carbon.mind.active_addictions, type, 1) //Keeps withdrawal at first cycle.
+	SEND_SIGNAL(affected_carbon, COMSIG_CLEAR_MOOD_EVENT, "[type]_addiction")
+
+/// Called when addiction is in stage 1 every process
+/datum/addiction/proc/withdrawal_stage_1_process(mob/living/carbon/affected_carbon, delta_time)
+	if(DT_PROB(5, delta_time))
+		to_chat(affected_carbon, "<span class='danger'>[withdrawal_stage_messages[1]]</span>")
```

Comment:
these do not check if the message is valid, so an empty list will cause errors

Reply:
Is a null list better in this scenario?

Reply:
I told you to add handling for an empty list in my original review, but it wasn't added. That would be fine although technically null rather than list() is more performant

Reply:
Sounds good

Reply:
d5e5da9

Reply:
I sweear to fucking god if you don't put a NULL CHECK on the list here i am going to explode

Reply:
the proc will now check if `addiction_stage_message == null` prior to trying to send a message.

sorry for the oversight

9289eec